feat: add deployment branch support

amustaque97 opened the pull request:
Closes: #2132
Extends environment configuration to support specifying deployment branch restrictions, allowing teams to control which branches can deploy to each environment via a new field in the TOML schema.

Kobzol commented on repos/rust-lang/edition-guide.toml:
```diff
@@ -12,6 +12,6 @@ pattern = "master"
-[[environments]]
-name = "github-pages"
+[environments.github-pages]
+branches = []
```
Just to avoid a common default case, if the branches array is empty, I'd remove it from the current files, and keep the default.

amustaque97 commented on sync-team/src/github/api/write.rs:
```diff
@@ -500,14 +507,184 @@ impl GitHubWrite {
+            } else {
+                serde_json::json!({
+                    "deployment_branch_policy": {
+                        "protected_branches": false,
```
I'm using api endpoint https://docs.github.com/en/rest/deployments/environments?apiVersion=2022-11-28#create-or-update-an-environment. I assume that we will deploy branches with patterns, so setting protected_branches as false

<img width="1512" height="982" alt="Screenshot 2025-12-05 at 12 37 58 PM" src="https://github.com/user-attachments/assets/2d2286e7-9fd3-405b-b8d7-ab5caf9de479" />

marcoieni commented on sync-team/src/github/mod.rs:
```diff
@@ -865,7 +878,7 @@ struct CreateRepoDiff {
     permissions: Vec<RepoPermissionAssignmentDiff>,
     branch_protections: Vec<(String, api::BranchProtection)>,
-    environments: Vec<String>,
+    environments: Vec<(String, Vec<String>)>,
```
I would substitute `Vec<String>` with a struct

amustaque97 replied:
I would like to change either whole  `Vec<(String, Vec<String>)>` as a struct, not just `Vec<String>`. IMO, it will make it unnecessarily complex. 
```rust
struct Branches {
    names: Vec<String>
}
```


I would like to keep it as it is. WDYT?

amustaque97 commented:
> From [#2159 (comment)](https://github.com/rust-lang/team/pull/2159#issuecomment-3617834360) you can update the toml files to import the branches 👍

I made the required changes 🙌🏻

Kobzol commented on rust_team_data/src/v1.rs:
```diff
@@ -175,7 +175,7 @@ pub struct Repo {
     pub branch_protections: Vec<BranchProtection>,
     pub crates: Vec<Crate>,
-    pub environments: Vec<Environment>,
+    pub environments: IndexMap<String, Environment>,
```
This is a breaking change. I think that we can leave the environments as they were in the v1 API, and just add a `branches` field to the environment.

Kobzol replied:
To clarify, the contents of the v1 API is used by several other projects. If this PR is merged as-is, those projects will try to deserialize this field as a `Vec` (because they will still have the old `serde` definition), but the API will now expose it as a JSON dictionary. So the repository endpoint will stop working for them.

amustaque97 replied:
I'm not sure. Is there any way to confirm this? As environment feature was implemented very recently https://github.com/rust-lang/team/pull/2127, if we still have time, I would like to use the current structure because IMO this is how the schema should be.

I would love to hear your thoughts @Kobzol

Kobzol replied:
Hmm, in theory if the projects haven't updated their `team` version *yet*, it would be doable. Here is a list of some projects that use the API: https://github.com/search?type=code&q=%22rust_team_data+%3D+%22+org%3Arust-lang We'd need to check what version of `team` they use, if they use the repos endpoint.

Kobzol commented on src/validate.rs:
```diff
@@ -1058,28 +1058,39 @@ fn validate_environments(data: &Data, errors: &mut Vec<String>) {
+                        env_name
+                    );
+                }
+                if !seen_branches.insert(branch) {
```
Is this really an issue? Why couldn't I have two environments on the same branch?

amustaque97 replied:
> Is this really an issue?

Not really, this is for better logging and checks. This will be helpful when we try to add the same branch twice or more in any environment. 

> Why couldn't I have two environments on the same branch?

It’s allowed — a single branch can belong to multiple environments. As you can see in the screenshot below, we iterate over each environment and check for duplicates within that scope.

<img width="834" height="739" alt="Screenshot 2025-12-11 at 5 25 12 PM" src="https://github.com/user-attachments/assets/3b0aa34f-fef3-47dc-a131-29591f41ce72" />

amustaque97 commented on sync-team/src/github/mod.rs:
```diff
@@ -1092,7 +1123,36 @@ impl std::fmt::Display for UpdateRepoDiff {
+                            writeln!(f, "        Branches: {}", branches.join(", "))?;
+                        }
+                    }
+                    EnvironmentDiff::Update(name, old_branches, new_branches) => {
```
I got you, and it 100% makes sense. I have made the required changes.

Kobzol commented:
Some repositories use tags for protection rules, not branches. This PR will (I think) set everything as branches. I wonder if that will work when a repository then tries to deploy from a tag, have you tried that? The type (branch/tag) is not mandatory when creating the branch policy, but I think that if it doesn't match, it might be an issue. So in that case we might want to explicitly model both branches and tags in `team`.

amustaque97 commented:
> Some repositories use tags for protection rules, not branches. This PR will (I think) set everything as branches. I wonder if that will work when a repository then tries to deploy from a tag. Have you tried that? The type (branch/tag) is not mandatory when creating the branch policy, but I think that if it doesn't match, it might be an issue. So in that case we might want to explicitly model both branches and tags in `team`.

Whoops 😬 , I implemented changes based on the comment https://github.com/rust-lang/team/issues/2132#issuecomment-3591469765; I have added support for tags, too.  Can you please review the PR again?

Kobzol commented:
I see, I didn't know that. It seemed in https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra/topic/mdbook.20trusted.20publishing.20broken/with/563418198 that tags would be required. I'll ask t-infra on the next meeting.

amustaque97 commented on docs/toml-schema.md:
```diff
@@ -433,16 +433,33 @@ merge-bots = ["homu"]
+## Legacy syntax (still supported for backwards compatibility):
+# The old "deployment-patterns" field is still supported
+[environments.legacy-patterns]
+deployment-patterns = [
```
whooops! This was a stupid mistake :( This was another implementation approach I tried and generated docs using LLM, and forgot to remove.

amustaque97 replied:
This is fixed now. Thank you again for highlighting it.
